File: nodegraph/p1_cpp_nodegraph/s2_dynamic_graph.cpp
#include<memory>
#include<set>
#include<iostream>

// local namespace
namespace {

/// @brief General node interface
class Node
{
public:
    virtual ~Node() {}

    Node() {}
    Node(Node const &) = delete;
    Node(Node &&) = delete;

    Node &operator= (Node const &) = delete;
    Node &operator= (Node &&) = delete;

    virtual void ProcessForwards() = 0;
    virtual void ProcessBackwards() = 0;

    virtual bool HasActiveInputs() = 0;
    virtual bool HasActiveOutputs() = 0;
};

/// @brief Universal pin
/// @tparam T Type of data stored in this pin
template<class T> class Pin
{
public:
    /// @brief Obtain owning node
    Node &GetOwningNode() { return m_node; }

    /// @brief Obtain connected pin
    Pin<T> *GetConnectedPin() { return m_connection; }

    /// @brief Set data stored in this pin
    template<class X> void SetData(X &&x) { m_data = std::forward<X>(x); }

    /// @brief Get data stored in this pin
    T const & GetData() const { return m_data; }

    /// @brief Disconnect other pin
    void Disconnect()
    {
        if (m_connection)
        {
            m_connection->m_connection = nullptr;
            m_connection = nullptr;
        }
    }

private:
    Node &m_node;
    Pin<T> *m_connection;
    T m_data;

protected:
    /// @brief Create pin attached to specific node
    Pin(Node &node): m_node(node), m_connection(nullptr), m_data{} {}

    /// @brief Connect pin to another pin
    void Connect(Pin<T> &pin)
    {
        m_connection = &pin;
        pin.m_connection = this;
    }
};

// Forward declaration of OutputPin
template<class T> struct OutputPin;

/// @brief Input pin
/// @tparam T Type of data input
template<class T> struct InputPin : Pin<T>
{
    InputPin(Node &node): Pin<T>(node) {}

    /// @brief Connect output pin (implemented below to solve catch22)
    void Connect(OutputPin<T> &pin);

    /// @brief Receive data from the other connected pin into this pin
    void ReceiveData()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            Pin<T>::SetData(connectedPin->GetData());
        }
    }
    
    /// @brief Propagate processing to connected node
    void PropagateBackwards()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->GetOwningNode().ProcessBackwards();
        }
    }
};

/// @brief Output pin
/// @tparam T Type of data output
template<class T> struct OutputPin : Pin<T>
{
    OutputPin(Node &node): Pin<T>(node) {}

    /// @brief Connect output pin
    void Connect(InputPin<T> &pin) { Pin<T>::Connect(pin); }

    /// @brief Send data from this pin to the other connected pin
    void SendData()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->SetData(Pin<T>::GetData());
        }
    }

    /// @brief Propagate processing to connected node
    void PropagateForwards()
    {
        if (auto connectedPin = Pin<T>::GetConnectedPin(); nullptr != connectedPin)
        {
            connectedPin->GetOwningNode().ProcessForwards();
        }
    }
};

// Implementation of InputPin::Connect
template<class T> void InputPin<T>::Connect(OutputPin<T> &pin) { Pin<T>::Connect(pin); }

/// @brief Node that will provide data
/// @tparam T Type of source data
template<class T> class SourceNode : public Node
{
public:
    ~SourceNode() {}
    SourceNode() : m_outputPin{*this}
    {}

    void ProcessForwards() override
    {
        GetOutputPin().SendData();
        GetOutputPin().PropagateForwards();
    }
    
    void ProcessBackwards() override
    {
        // This could be loading new value from somewhere
    }
    
    bool HasActiveInputs() override { return false; }
    bool HasActiveOutputs() override { return true; }

    template<class X> void LoadValue(X &&x)
    {
        GetOutputPin().SetData(std::forward<X>(x));
    }

    OutputPin<T> &GetOutputPin() { return m_outputPin; }

private:
    OutputPin<T> m_outputPin;
};

/// @brief Node that will receive final result of all procesing
/// @tparam T Type of final result data
template<class T> class TargetNode : public Node
{
public:
    ~TargetNode() {}
    TargetNode() : m_inputPin{*this}
    {}

    void ProcessForwards() override
    {
        // This could render result somewhere
    }

    void ProcessBackwards() override
    {
        GetInputPin().PropagateBackwards();
        GetInputPin().ReceiveData();
    }
    
    bool HasActiveInputs() override { return true; }
    bool HasActiveOutputs() override { return false; }

    const T &GetValue() const 
    {
        // Process(); -- should we pull?
        return GetInputPin().GetData();
    }

    InputPin<T> &GetInputPin() { return m_inputPin; }
    InputPin<T> const &GetInputPin() const { return m_inputPin; }

private:
    InputPin<T> m_inputPin;
};

template<class X, class Y, class F> class TransformNode : public Node
{
public:
    ~TransformNode() {}
    TransformNode(F f): m_function(std::move(f)), m_inputPin{*this}, m_outputPin{*this} {}
    
    void ProcessBackwards() override
    {
        // If we're propagating backwards, then request came from output pin
        // This can happen when we're pulling from output
        GetInputPin().PropagateBackwards();
        GetInputPin().ReceiveData();
        GetOutputPin().SetData(m_function(GetInputPin().GetData()));
    }

    void ProcessForwards() override
    {
        // If we're propagate forwards, then request came from input pin
        // This can happen when we're pushing new input data
        GetOutputPin().SetData(m_function(GetInputPin().GetData()));
        GetOutputPin().SendData();
        GetOutputPin().PropagateForwards();
    }
    
    bool HasActiveInputs() override { return true; }
    bool HasActiveOutputs() override { return true; }

    InputPin<X> &GetInputPin() { return m_inputPin; }
    OutputPin<Y> &GetOutputPin() { return m_outputPin; }

private:
    F m_function;
    InputPin<X> m_inputPin;
    OutputPin<Y> m_outputPin;
};

class NodeGraph
{
public:
    template<class NodeT>
    NodeGraph &AddNode(std::unique_ptr<NodeT> &&nodePtr)
    {
        m_nodes.emplace(std::move(nodePtr));
        return *this;
    }

    std::set<Node *> GetSourceNodes()
    {
        std::set<Node *> sourceNodes{};

        // One could use combination of std::transform() and std::copy_if(), or std::ranges
        for (auto &nodePtr : m_nodes)
        {
            if (not nodePtr->HasActiveInputs()) { sourceNodes.insert(nodePtr.get()); }
        }
    
        return std::move(sourceNodes);
    }

    std::set<Node *> GetTargetNodes()
    {
        std::set<Node *> targetNodes{};

        for (auto &nodePtr : m_nodes)
        {
            if (not nodePtr->HasActiveOutputs()) { targetNodes.insert(nodePtr.get()); }
        }

        return std::move(targetNodes);
    }

private:
    std::set<std::unique_ptr<Node>> m_nodes;
};

struct ExampleDataSample
{
    ExampleDataSample(): x{}, y{} {}
    ExampleDataSample(int ix, int iy): x{ix}, y{iy} {}

    int x;
    int y;
};

struct ExampleDataResult
{
    ExampleDataResult(): u{}, v{} {}
    ExampleDataResult(double iu, double iv): u{iu}, v{iv} {}

    double u;
    double v;
};

std::ostream &operator <<(std::ostream &os, ExampleDataSample const &data)
{
    return os << "[ x: " << data.x << ", y: " << data.y << "]";
}

std::ostream &operator <<(std::ostream &os, ExampleDataResult const &data)
{
    return os << "[ u: " << data.u << ", v: " << data.v << "]";
}

std::pair<
    SourceNode<std::shared_ptr<ExampleDataSample>> &,
    TargetNode<std::shared_ptr<ExampleDataResult>> &> add_example_nodes_to_graph(NodeGraph &graph)
{
    // Func lives only within current scope.
    // Note that it gets consumed by TrasformNode constructor.
    auto func = [](std::shared_ptr<ExampleDataSample> const &data)
    {
        return std::make_shared<ExampleDataResult>(
            static_cast<double>(data->x * data->y),
            static_cast<double>(data->x) / static_cast<double>(data->y));
    };

    // Let's create node using our transformation function
    auto transformNodePtr = std::make_unique<TransformNode<
        std::shared_ptr<ExampleDataSample>,
        std::shared_ptr<ExampleDataResult>,
        decltype(func)>>(func);

    // Now we need some source
    auto sourceNodePtr = std::make_unique<SourceNode<
        std::shared_ptr<ExampleDataSample>>>();

    // And we also need some target
    auto targetNodePtr = std::make_unique<TargetNode<
        std::shared_ptr<ExampleDataResult>>>();

    transformNodePtr->GetInputPin().Connect(sourceNodePtr->GetOutputPin());
    transformNodePtr->GetOutputPin().Connect(targetNodePtr->GetInputPin());

    auto &sourceNode = *sourceNodePtr;
    auto &targetNode = *targetNodePtr;

    // Move ownership of the nodes into node graph
    graph
        .AddNode(std::move(transformNodePtr))
        .AddNode(std::move(sourceNodePtr))
        .AddNode(std::move(targetNodePtr));

    // WARNING !!! At this point unique node pointers are no longer valid

    return {sourceNode, targetNode};
}

} // end of local namespace

void test_with_node_graph()
{
    NodeGraph graph{};

    const auto [sourceNode, targetNode] = add_example_nodes_to_graph(graph);

    // Here we will push the value from the source
    sourceNode.LoadValue(std::make_shared<ExampleDataSample>(7, 9));
    sourceNode.ProcessForwards();
    
    std::cout << "Result of processing forwards: f(" << *sourceNode.GetOutputPin().GetData() << ") => " << *targetNode.GetValue() << std::endl;
    
    // Here we will pull the value from the target
    sourceNode.LoadValue(std::make_shared<ExampleDataSample>(11, 5));
    targetNode.ProcessBackwards();
    
    std::cout << "Result of processing backwards: f(" << *sourceNode.GetOutputPin().GetData() << ") => " << *targetNode.GetValue() << std::endl;
}

File: nodegraph/p2_rust_nodegraph/src/s2_dynamic_graph.rs
use std::cell::RefCell;


trait GraphNode<'a>
{
    fn process_forwards(&self);
    fn process_backwards(&self);
    fn setup(&'a self);
}

#[derive(Default)]
struct GraphNodePin<'a, T> where T: Clone + Default
{
    node: RefCell<Option<&'a dyn GraphNode<'a>>>,
    connection: RefCell<Option<&'a GraphNodePin<'a, T>>>,
    data: RefCell<T>
}

impl<'a, T> GraphNodePin<'a, T> where T: Clone + Default
{
    pub fn get_node(&self) -> Option<&'a dyn GraphNode> {
        self.node.borrow().clone()
    }

    pub fn get_connection(&self) -> Option<&'a GraphNodePin<'a, T>> {
        self.connection.borrow().clone()
    }

    pub fn setup(&self, node: &'a dyn GraphNode<'a>) {
        self.node.replace(node.into());
    }

    pub fn connect(&self, pin: &'a GraphNodePin<'a, T>) {
        self.connection.replace(pin.into());
    }

    pub fn set_data(&self, data: T) {
        self.data.replace(data);
    }

    pub fn get_data(&self) -> T {
        self.data.borrow().clone()
    }
}

#[derive(Default)]
struct GraphNodeInputPin<'a, T> where T: Clone + Default
{
    pin: GraphNodePin<'a, T>
}

impl<'a, T> GraphNodeInputPin<'a, T> where T: Clone + Default
{
    pub fn connect(&'a self, output_pin: &'a GraphNodeOutputPin<'a, T>) {
        self.pin.connect(&output_pin.pin);
        output_pin.pin.connect(&self.pin);
    }

    pub fn receive_data(&self) {
        if let Some(input) = self.pin.get_connection() {
            self.pin.set_data(input.get_data());
        }
    }

    pub fn propagate_backwards(&self) {
        if let Some(input) = self.pin.get_connection() {
            if let Some(node) = input.get_node() {
                node.process_backwards();
            }
        }
    }
}

#[derive(Default)]
struct GraphNodeOutputPin<'a, T> where T: Clone + Default
{
    pin: GraphNodePin<'a, T>
}

impl<'a, T> GraphNodeOutputPin<'a, T> where T: Clone + Default
{
    pub fn connect(&'a self, input_pin: &'a GraphNodeInputPin<'a, T>) {
        self.pin.connect(&input_pin.pin);
        input_pin.pin.connect(&self.pin);
    }

    pub fn send_data(&self) {
        if let Some(output) = self.pin.get_connection() {
            output.set_data(self.pin.get_data());
        }
    }
    
    pub fn propagate_forwards(&self) {
        if let Some(output) = self.pin.get_connection() {
            if let Some(node) = output.get_node() {
                node.process_forwards();
            }
        }
    }
}

#[derive(Default)]
struct SourceGraphNode<'a, T> where T: Clone + Default
{
    pub output_pin: GraphNodeOutputPin<'a, T>
}

impl<'a, T> SourceGraphNode<'a, T> where T: Clone + Default
{
    pub fn load_value(&self, data: T)
    {
        self.output_pin.pin.set_data(data);
    }
}

impl<'a, T> GraphNode<'a> for SourceGraphNode<'a, T> where T: Clone + Default
{
    fn process_backwards(&self) {
    }

    fn process_forwards(&self) {
        self.output_pin.send_data();
        self.output_pin.propagate_forwards();
    }

    fn setup(&'a self) {
        self.output_pin.pin.setup(self);
    }
}

#[derive(Default)]
struct TargetGraphNode<'a, T> where T: Clone + Default
{
    pub input_pin: GraphNodeInputPin<'a, T>
}

impl<'a, T> TargetGraphNode<'a, T> where T: Clone + Default
{
    pub fn get_value(&self) -> T
    {
        self.input_pin.pin.get_data()
    }
}

impl<'a, T> GraphNode<'a> for TargetGraphNode<'a, T> where T: Clone + Default
{
    fn process_backwards(&self) {
        self.input_pin.propagate_backwards();
        self.input_pin.receive_data();
    }

    fn process_forwards(&self) {
    }
    
    fn setup(&'a self) {
        self.input_pin.pin.setup(self);
    }
}

struct TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    pub input_pin: GraphNodeInputPin<'a, X>,
    pub output_pin: GraphNodeOutputPin<'a, Y>,
    pub func: F
}

impl<'a, X, Y, F> TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    pub fn new(func: F) -> Self {
        Self {
            input_pin: Default::default(),
            output_pin: Default::default(),
            func
        }
    }
}

impl<'a, X, Y, F> GraphNode<'a> for TransformGraphNode<'a, X, Y, F>
where
    X: Clone + Default,
    Y: Clone + Default,
    F: Fn(X) -> Y
{
    fn process_backwards(&self) {
        self.input_pin.propagate_backwards();
        self.input_pin.receive_data();
        self.output_pin.pin.set_data((&self.func)(self.input_pin.pin.get_data()));
    }

    fn process_forwards(&self) {
        self.output_pin.pin.set_data((&self.func)(self.input_pin.pin.get_data()));
        self.output_pin.send_data();
        self.output_pin.propagate_forwards();
    }
    
    fn setup(&'a self) {
        self.input_pin.pin.setup(self);
        self.output_pin.pin.setup(self);
    }
}

struct NodeGraph<'a> {
    nodes: Vec<Box<dyn GraphNode<'a> + 'a>>
}

#[test]
fn test_me() {
    let foo = |x: i32| -> i32 { x * x };

    let source_node = SourceGraphNode::<'_, i32>::default();
    let target_node = TargetGraphNode::<'_, i32>::default();
    let transform_node = TransformGraphNode::new(foo);

    let graph = NodeGraph {
        nodes: vec![
            Box::new(source_node),
            Box::new(target_node),
            Box::new(transform_node)
        ]
    };

    // We cannot turn static NodeGraph with lifetime references into dynamic
    // graph by boxing the nodes. This is because boxing nodes would allow nodes
    // to live different lifetime and that prohibits cyclic references created
    // when connecting nodes, and self-references when we box the nodes.
    
    //(&graph.nodes[0]).setup();
}
